3.x - Strip aliases from update query conditions where possible.

ndm2 opened the pull request:
Unlike for delete queries, possible aliases in the conditions of update queries are not being removed, which will fail for some database dialects. Given that update queries currently do not support joins, just like delete queries, removing the aliases here in the same way should be good enough for now I think.

The fact that expressions are being left untouched is something that maybe should go in the docs somewhere, it's something that affects all dependent, non-cascading associated deletes as well, which developers should generally be aware of.

refs #9404

Commit:
Strip aliases from update query conditions where possible.

Unlike for delete queries, possible aliases in the conditions of
update queries are not being removed, which will fail for some
database dialects.

thinkingmedia commented:
Which database dialects is this an issue for?

ndm2 commented:
@thinkingmedia For MySQL and Postgres it seems, at least they failed on Travis.

* **https://travis-ci.org/cakephp/cakephp/builds/157474410**
* **https://travis-ci.org/cakephp/cakephp/jobs/157474412#L699**

As mentioned in #9404, except Sqlite, all supported dialects seem to generally support aliases in update queries, but at least SQL server, and possibly Postgres require an additional `FROM`.

Adding alias support for dialects that support it would of course be an option too, but it will require larger changes, and possibly also API changes with regards to how `Query::update()` works, which currently accepts a string (though it say the argument is unused), and doesn't take `from()` into account at all.

thinkingmedia commented:
@ndm2 yes, but I ran into this issue before.

The aliases are only added to the where clause when you create the query object improperly for an UPDATE statement.

I don't know the exact conditions that raise this issue off the top of my head. Let me go back to my code and see where I ran into this before. I'll be back.

ndm2 commented:
In the case of #9404 it happens indirectly for conditions in non-dependent/cascading hasMany associations using the replace save strategy. The possibly aliased conditions will be passed to a `Table::updateAll()` call.

**https://github.com/ndm2/cakephp/blob/907df7743bc58494273b72bcd3c128a206fb1c7b/src/ORM/Association/HasMany.php#L471-L472**

thinkingmedia commented:
I thought (and I could be remembering incorrectly) that there was a flag or something in the Query object to indicate if the where clause should be aliased. This is why the `updateAll` method does not have aliases on the conditions.

thinkingmedia commented:
Yes, I see what's happening now. Update/Delete use conditions as-is, but a SELECT query will call `aliasFields` in QueryTrait.

I suspect this is an issue, because the condition clauses where copied from a SELECT query to be used with a DELETE/UPDATE query.

Am I correct?

I fear there is no easy fix this. I know this PR will going to break joins on DELETE/UPDATE later. Can you add a future proof check to see if the query contains the join clause. If so, maybe throw an exception or trigger_error. That way, in the future when someone wants to add joins they will see what's going on here.

ndm2 commented:
> I suspect this is an issue, because the condition clauses where copied from a SELECT query to be used with a DELETE/UPDATE query.

No, there's no select involved in case of #9404, it's just association conditions being passed as is to `where()` in a `update()->set()->where()` call via `Table::updateAll()`.


> That way, in the future when someone wants to add joins they will see what's going on here.

Are you referring to when someone wants to implement join support for `UPDATE`/`DELETE` queries in the future? Surely the code removing the aliases could trigger a warning/exception in case the query has any joins.

lorenzo commented on src/Database/SqlDialectTrait.php:
```diff
@@ -178,6 +178,34 @@ protected function _deleteQueryTranslator($query)
+     * @param \Cake\Database\Query $query The query to process.
+     * @return \Cake\Database\Query The modified query.
+     */
+    protected function _removeAliasesFromConditions($query)
```
This just hit me now... what happens in the case of having a subquery in conditions? Are we stripping the aliases as well?

ndm2 replied:
I don't think so, `Query` implements `ExpressionInterface`, so they shouldn't make it past the instance check.

lorenzo replied:
Yeah, but the travers will keep going down all leaves inside an ExpressionInterface

ndm2 replied:
Ah OK, I see... a quick test shows that it doesn't, which I guess is because unline `QueryExpression::traverse()`, `Query::traverse()` isn't recursive, so for queries it will just pass the parts to the visotor, which will discard them at the `instanceof Comparison` check.

I'll have a closer look later on and add some subqueries to the test.

Commit:
Test subqueries in conditions too.

Commit:
Trigger a notice when removing aliases from queries with joins.

Removing aliases from query conditions can break references to
joins. Since joins for delete/update queries aren't yet supported,
this notice is ment to serve as a hint for future implementations.

jippi commented on src/Database/SqlDialectTrait.php:
```diff
@@ -206,6 +206,14 @@ protected function _updateQueryTranslator($query)
     protected function _removeAliasesFromConditions($query)
     {
+        if (!empty($query->clause('join'))) {
+            trigger_error(
```
i can haz exception or nothing - E_USER_NOTICE would just be annoying from monitoring perspective in newrelic during production :)

ndm2 replied:
@jippi I've settled for a notice as I thought introducing exceptions in patch releases would be discouraged?, and because the only purpose of this notice is to inform developers that are trying to implement join support.

ADmad replied:
Introducing exceptions in bugfix  release has caused  problems in the past.

thinkingmedia replied:
@ADmad yes, I understand your concern, but this is an edge case that won't happen. The ORM currently does not support JOINs on UPDATE/DELETE. This error is to future proof changes so that when they are added this code is updated.

lorenzo replied:
I vote for exception as well

ndm2 replied:
Exception it is then...

Commit:
Throw an exception instead of triggering a notice.

This situation is unlikely to happen currently, given that joins
aren't supported for delete/update queries, therefore introducing
an exception at this point should be safe.

ndm2 commented:
Are there any further concerns?

dereuromark commented on src/Database/SqlDialectTrait.php:
```diff
@@ -178,6 +178,43 @@ protected function _deleteQueryTranslator($query)
+     */
+    protected function _removeAliasesFromConditions($query)
+    {
+        if (!empty($query->clause('join'))) {
```
This should not call a function inside empty().
Please use

    if ($query->clause(...))

markstory replied:
Why? It 'works' in PHP5.5+

ndm2 replied:
I'll probably never learn it, I'm using `empty()` way too often :) I don't see a general problem with `empty()` and function calls, but it surely is superfluos, as it's the same result without it - I'll remove it.

markstory commented on src/Database/SqlDialectTrait.php:
```diff
@@ -178,6 +178,43 @@ protected function _deleteQueryTranslator($query)
+    protected function _removeAliasesFromConditions($query)
+    {
+        if (!empty($query->clause('join'))) {
+            throw new \RuntimeException(
```
Won't this cause any and all `UPDATE` queries that contain joins to fail? Did those ever work before?

ndm2 replied:
They never worked, the compiler doesn't include the `joins` part.

**https://github.com/cakephp/cakephp/blob/3.3.7/src/Database/QueryCompiler.php#L60**

dereuromark commented:
If we are sure this won't be BC breaking, this looks good for master then.
Afterwise, to be sure we could also target 3.4 and merge there.

Commit:
Do not use `empty()`, it's not necessary.

thinkingmedia commented:
I added this to a ORM project for 3.4. If it's still going to master just let me know, and I'll remove it.

dereuromark commented:
Please better rebase to 3.4.

ndm2 commented:
Whetever is deemed appropriate... I'd surely like to see it merged as soon as possible, as #9004 is relly annoying.

thinkingmedia commented:
@ndm2 I think the 3.3 branch has entered into a stability phase now. Can you change the base to `3.next`
